File: backend/app/core/upgrades/stat_modifiers/bullet_damage_1.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class BulletDamage1(UpgradeImplementation):
    id = 'bullet_damage_1'
    name = 'Sharper Rounds'
    description = '+1 bullet damage'
    rarity = 'common'
    type = 'stat_modifier'
    cost = 2

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 1
        appliers.modifiers.add_multiplier('attack', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/bullet_damage_1.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID, AttackTypeID } from '../../data/ID'

export const BulletDamage1Def: UpgradeDef = {
  id: "bullet_damage_1",
  name: "Sharper Rounds",
  description: "+1 bullet damage",
  rarity: RarityID.Common,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 2,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: 1,
  isMultiplier: false,
  specificAttackType: AttackTypeID.Bullet,
  stackable: true,
  maxStacks: 99999,
}

export class BulletDamage1 extends Upgrade {}
